content/doc/developer/testing/unit-test.adoc

```diff
@@ -0,0 +1,541 @@
+
+When a test completes, the temporary Jenkins installation will be
+destroyed.
+
+[source,groovy]
+----
+import org.jvnet.hudson.test.JenkinsRule;
+import org.apache.commons.io.FileUtils;
+import hudson.model.*;
+import hudson.tasks.Shell;
+import org.junit.Test;
+import org.junit.Rule;
+public class AppTest {
+  @Rule public JenkinsRule j = new JenkinsRule();
+  @Test public void first() throws Exception {
+    FreeStyleProject project = j.createFreeStyleProject();
+    project.getBuildersList().add(new Shell("echo hello"));
+    FreeStyleBuild build = project.scheduleBuild2(0).get();
+    System.out.println(build.getDisplayName() + " completed");
+    // TODO: change this to use HtmlUnit
+    String s = FileUtils.readFileToString(build.getLogFile());
+    assertThat(s, contains("+ echo hello"));
+  }
+}
+----
+
+Tip
+
+You can use `+@ClassRule+` to speed up the test suite if you are
+_positive_ that the test cases will not interfere with each other.
+
+[[UnitTest-Howtosetenvironmentvariables]]
+=== How to set environment variables
+
+When preparing your virtual test environment, you may wish to simulate
+Jenkins environment variables that can be set on the Jenkins
+configuration page. Adding environment variables to a Jenkins instance
+before a test is simple, as the example below demonstrates.
+
+[source,groovy]
+----
+public void setEnvironmentVariables() throws IOException {
+    EnvironmentVariablesNodeProperty prop = new EnvironmentVariablesNodeProperty();
+    EnvVars envVars = prop.getEnvVars();
+    envVars.put("sampleEnvVarKey", "sampleEnvVarValue");
+    j.jenkins.getGlobalNodeProperties().add(prop);
+}
+----
+
+[[UnitTest-Debugging]]
+== Debugging
+
+Your IDE will most likely have the ability to select a single JUnit test
+and execute it in the debugger. Otherwise you can run
+`+mvn -Dmaven.surefire.debug -Dtest=hudson.SomeTest test+` to accomplish
+this.
+
+To debug agents launched by Jenkins, set
+`+-Dorg.jvnet.hudson.test.HudsonTestCase.slaveDebugPort=PORT+` to the
+system property, or from your test case set
+`+JenkinsRule.SLAVE_DEBUG_PORT+` to a non-0 value.
+
+[[UnitTest-VariousTestTechniques]]
+== Various Test Techniques
+
+[[UnitTest-Stubbing]]
+=== Stubbing
+
+Sometimes you want to have quick tests which don't start up a 'full'
+Jenkins instance - as `+JenkinsRule+` does - as this can take some time.
+In that case you shouldn't have your test classes use `+JenkinsRule+`.
+(Or you can use `+@WithoutJenkins+` on certain methods.)
+
+As creating most Jenkins core classes without a Jenkins instance is
+unfortunately not easy, _stubbing_ can come in handy. One excellent
+stubbing and mocking framework is e.g. http://mockito.org/[Mockito]. For
+example, if you want to stub a build with a certain result you could do:
+
+[source,groovy]
+----
+AbstractBuild build = Mockito.mock(AbstractBuild.class);
+Mockito.when(build.getResult()).thenReturn(Result.FAILURE);
+----
+
+See
+
+<<Mocking-in-Unit-Tests>>
```

Comment:
Unless you've created a heading named "Mocking in Unit Tests", you can't reference it with `<<Mocking-in-Unit-Tests>>`
```suggestion
```